honeybadgermpc/protocols/avid.py

```
@@ -0,0 +1,225 @@
+        self.ok_queue = asyncio.Queue()
+
+    def broadcast(self, o):
+        for i in range(self.n):
+            self.send(i, o)
+
+    async def retrieve(self, sid, index):
+        """ Retrieve individual item
+        :param int index:  the index of retrieval
+        """
+        # we are able to retrieve things from AVID indivdually
+        assert 0 <= index < self.input_size
+
+        # send retrieve until ok state
+        assert await self.ok_queue.get()
+
+        # send retrieve to all the parties
+        self.broadcast((sid, AVIDMessageType.RETRIEVE, index))
+
+        result = [None] * self.input_size
+        response_set = set()
+
+        # response threshold same as k
+        response_threshold = self.n - 2 * self.t
+
+        def decode_output(roothash):
+            decoded_output = decode(response_threshold, self.n, result)
+            # Rebuild the merkle tree to guarantee decoding is correct
+            _stripes = encode(response_threshold, self.n, decoded_output)
+            _mt = merkle_tree(_stripes)
+            _roothash = _mt[1]
+            try:
+                assert _roothash == roothash
+            except Exception as e:
+                logger.error("Failed to verify merkle tree: %s", e)
+            return decoded_output
+
+        while True:  # recv loop for retrieve
+            sender, msg = await self.retrieval_queue.get()
+            if msg[1] == AVIDMessageType.RESPONSE:
+                (_, _, roothash, data) = msg
+
+                if sender in response_set:
+                    logger.warning("Redundant RESPONSE from %s", sender)
+                    continue
+
+                result[sender] = data
+                response_set.add(sender)
+
+                if len(response_set) >= response_threshold:
+                    m = ""
+                    try:
+                        m = decode_output(roothash)
+                    except Exception as e:
```

You should catch only `HoneyBadgerMPCError` or the error that you throw. Don't catch all exceptions. See if reliablebroadcast.py defines a separate exception when merkle tree verification fails.

---

reliablebroadcast doesn't do anything.

---

No, you should not eat up all exceptions. This is not good practice. You should only catch `HoneyBadgerMPCError`.

---

But decode doesn't throw ``HoneyBadgerMPCError ``

---

What is the rule for HoneyBadgerMPCError? We should have something we throw that is caused by clearly incorrect messages sent from another peer, and that should be the only exception we catch here. An